<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create Arrow Files — createArrowFiles • ArchR</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create Arrow Files — createArrowFiles" />
<meta property="og:description" content="This function will create ArrowFiles from input files. These ArrowFiles are the main constituent for downstream analysis in ArchR." />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ArchR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/Articles/tutorial.html">Tutorial</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Articles/doubletRemoval.html">Doublet Removal</a>
    </li>
    <li>
      <a href="../articles/Articles/qcFilters.html">QC Filters</a>
    </li>
    <li>
      <a href="../articles/Articles/peakCalling.html">Peak Calling</a>
    </li>
    <li>
      <a href="../articles/Articles/dimReduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/Articles/geneScores.html">Gene Activity Scores</a>
    </li>
    <li>
      <a href="../articles/Articles/markerFeatures.html">Marker Feature Identification</a>
    </li>
    <li>
      <a href="../articles/Articles/signalTracks.html">Signal Track Ploting</a>
    </li>
    <li>
      <a href="../articles/Articles/footprinting.html">TF Footprinting</a>
    </li>
    <li>
      <a href="../articles/Articles/pseudobulkGeneration.html">Pseudobulk Replicate Generation</a>
    </li>
    <li>
      <a href="../articles/Articles/annotations.html">Gene and Genome Annotation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/Articles/faq.html">FAQ</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GreenleafLab/ArchR">
    <span class="fa fa-github"></span>
     
    ArchR on GitHub
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create Arrow Files</h1>
    
    <div class="hidden name"><code>createArrowFiles.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function will create ArrowFiles from input files. These ArrowFiles are the main constituent for downstream analysis in ArchR.</p>
    </div>

    <pre class="usage"><span class='fu'>createArrowFiles</span>(<span class='kw'>inputFiles</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sampleNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>outputNames</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"./"</span>, <span class='no'>sampleNames</span>), <span class='kw'>validBarcodes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>geneAnnotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genomeAnnotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>filterFrags</span> <span class='kw'>=</span> <span class='fl'>1000</span>,
  <span class='kw'>filterTSS</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>removeFilteredCells</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>minFrags</span> <span class='kw'>=</span> <span class='fl'>500</span>,
  <span class='kw'>outDir</span> <span class='kw'>=</span> <span class='st'>"QualityControl"</span>, <span class='kw'>nucLength</span> <span class='kw'>=</span> <span class='fl'>147</span>, <span class='kw'>TSSParams</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(),
  <span class='kw'>excludeChr</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"chrM"</span>, <span class='st'>"chrY"</span>), <span class='kw'>nChunk</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>bcTag</span> <span class='kw'>=</span> <span class='st'>"qname"</span>,
  <span class='kw'>gsubExpression</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>bamFlag</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>offsetPlus</span> <span class='kw'>=</span> <span class='fl'>4</span>,
  <span class='kw'>offsetMinus</span> <span class='kw'>=</span> -<span class='fl'>5</span>, <span class='kw'>addTileMat</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>TileMatParams</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(),
  <span class='kw'>addGeneScoreMat</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>GeneScoreMatParams</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(), <span class='kw'>force</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fu'><a href='getArchRThreads.html'>getArchRThreads</a></span>(), <span class='kw'>parallelParam</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>verboseHeader</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verboseAll</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>inputFiles</th>
      <td><p>A character vector containing the names of the input files to use to generate the Arrowfiles. These files can be  either scATAC tabix fragment files or a bam file.</p></td>
    </tr>
    <tr>
      <th>sampleNames</th>
      <td><p>A character vector containing the names to assign to the samples that correspond to the <code>inputFiles</code>. Each input file should receive a unique sample name. This list should be in the same order as <code>inputFiles</code>.</p></td>
    </tr>
    <tr>
      <th>outputNames</th>
      <td><p>The prefix to use for output files. Each input file should receive a unique output file name. This list should be in the same order as "inputFiles". For example, if the predix is "PBMC" the output file will be named "PBMC.arrow"</p></td>
    </tr>
    <tr>
      <th>validBarcodes</th>
      <td><p>A list of valid barcode strings to be used for filtering cells read from each input file (see <code><a href='getValidBarcodes.html'>getValidBarcodes()</a></code> for 10x fragment files).</p></td>
    </tr>
    <tr>
      <th>geneAnnotation</th>
      <td><p>The geneAnnotation (see <code>createGeneAnnotation()</code>) to associate with the ArrowFiles. This is used downstream to calculate TSS Enrichment Scores etc.</p></td>
    </tr>
    <tr>
      <th>genomeAnnotation</th>
      <td><p>The genomeAnnotation (see <code><a href='createGenomeAnnotation.html'>createGenomeAnnotation()</a></code>) to associate with the ArrowFiles. This is used downstream to collect chromosome sizes and nucleotide information etc.</p></td>
    </tr>
    <tr>
      <th>filterFrags</th>
      <td><p>The minimum number of mapped ATAC-seq fragments required per cell to pass filtering for use in downstream analyses. Cells containing greater than or equal to <code>filterFrags</code> total fragments wll be retained.</p></td>
    </tr>
    <tr>
      <th>filterTSS</th>
      <td><p>The minimum numeric transcription start site (TSS) enrichment score required for a cell to pass filtering for use in downstream analyses. Cells with a TSS enrichment score greater than or equal to <code>filterTSS</code> will be retained. TSS enrichment score is a measurement of signal-to-background in ATAC-seq.</p></td>
    </tr>
    <tr>
      <th>removeFilteredCells</th>
      <td><p>A boolean value that determines whether cells that do not pass <code>filterFrags</code> and <code>filterTSS</code> should be excluded entirely from the ArrowFiles. If <code>FALSE</code> cells that do not pass QC filters will be included in the ArrowFile but will be marked as not passing QC and excluded from downstream analyses.</p></td>
    </tr>
    <tr>
      <th>minFrags</th>
      <td><p>The minimum fragments per cell to be filtered immediately before any QC calculations (such as TSS Enrichment Score). This is useful for limiting the number of barcodes analyzed.</p></td>
    </tr>
    <tr>
      <th>outDir</th>
      <td><p>The relative path to the output directory for QC-level information and plots for each sample/ArrowFile.</p></td>
    </tr>
    <tr>
      <th>nucLength</th>
      <td><p>The length in basepairs that wraps around a nucleosome. This number is used for identifying fragments as sub-nucleosome-spanning, mono-nucleosome-spanning, or multi-nucleosome-spanning.</p></td>
    </tr>
    <tr>
      <th>TSSParams</th>
      <td><p>A list of parameters for computing TSS Enrichment scores.
This includes the <code>window</code> which is the size in basepairs of the window centered at each TSS (default 101),
the <code>flank</code> which is the size in basepairs of the flanking window (default 2000),
and the <code>norm</code> which describes the size in basepairs of the flank window to be used for normalization of the TSS enrichment score (default 100).
For example, given <code>window = 101, flank = 2000, norm = 100</code>, the accessibility within the 101-bp surrounding the TSS will be normalized to the accessibility in the 100-bp bins from -2000 bp to -1901 bp and 1901:2000.</p></td>
    </tr>
    <tr>
      <th>excludeChr</th>
      <td><p>A character vector containing the names of chromosomes to be excluded from downstream analyses. In most human/mouse analyses, this includes the mitochondrial DNA (chrM) and the male sex chromosome (chrY). This does, however, not exclude the corresponding fragments from being stored in the ArrowFile.</p></td>
    </tr>
    <tr>
      <th>nChunk</th>
      <td><p>The number of chunks to divide each chromosome into to allow for low-memory parallelized reading of the <code>inputFiles</code>. Higher numbers reduce memory usage but increase compute time.</p></td>
    </tr>
    <tr>
      <th>bcTag</th>
      <td><p>The name of the field in the input bam file containing the barcode tag information. See <code>ScanBam</code> in Rsamtools.</p></td>
    </tr>
    <tr>
      <th>gsubExpression</th>
      <td><p>A regular expression used to clean up the barcode tag string read in from a bam file. For example, if the barcode is appended to the readname or qname field like for the mouse atlas data from Cusanovic* and Hill* et al. (2018), the gsubExpression would be ":.*". This would retrieve the string after the colon as the barcode.</p></td>
    </tr>
    <tr>
      <th>bamFlag</th>
      <td><p>A vector of bam flags to be used for reading in fragments from input bam files. Should be in the format of a <code>scanBamFlag</code> passed to <code>ScanBam</code> in Rsamtools.</p></td>
    </tr>
    <tr>
      <th>offsetPlus</th>
      <td><p>The numeric offset to apply to a "+" stranded Tn5 insertion to account for the precise Tn5 binding site. See Buenrostro et al. Nature Methods 2013.</p></td>
    </tr>
    <tr>
      <th>offsetMinus</th>
      <td><p>The numeric offset to apply to a "-" stranded Tn5 insertion to account for the precise Tn5 binding site. See Buenrostro et al. Nature Methods 2013.</p></td>
    </tr>
    <tr>
      <th>addTileMat</th>
      <td><p>A boolean value indicating whether to add a "Tile Matrix" to each ArrowFile. A Tile Matrix is a counts matrix that, instead of using peaks, uses a fixed-width sliding window of bins across the whole genome. This matrix can be used in many downstream ArchR operations.</p></td>
    </tr>
    <tr>
      <th>TileMatParams</th>
      <td><p>A list of parameters to pass to the <code><a href='addTileMatrix.html'>addTileMatrix()</a></code> function. See <code><a href='addTileMatrix.html'>ArchR::addTileMatrix()</a></code> for options.</p></td>
    </tr>
    <tr>
      <th>addGeneScoreMat</th>
      <td><p>A boolean value indicating whether to add a Gene-Score Matrix to each ArrowFile. A Gene-Score Matrix uses ATAC-seq signal proximal to the TSS to estimate gene activity.</p></td>
    </tr>
    <tr>
      <th>GeneScoreMatParams</th>
      <td><p>A list of parameters to pass to the <code><a href='addGeneScoreMatrix.html'>addGeneScoreMatrix()</a></code> function. See <code><a href='addGeneScoreMatrix.html'>addGeneScoreMatrix()</a></code> for options.</p></td>
    </tr>
    <tr>
      <th>force</th>
      <td><p>A boolean value indicating whether to force ArrowFiles to be overwritten if they already exist in <code>outDir</code>.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>The number of threads to be used for parallel computing.</p></td>
    </tr>
    <tr>
      <th>parallelParam</th>
      <td><p>A list of parameters to be passed for biocparallel/batchtools parallel computing.</p></td>
    </tr>
    <tr>
      <th>verboseHeader</th>
      <td><p>A boolean value that determines whether standard output should include verbose sections.</p></td>
    </tr>
    <tr>
      <th>verboseAll</th>
      <td><p>A boolean value that determines whether standard output should include verbose subsections.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional args</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://twitter.com/JeffreyGranja'>Jeffrey Granja</a>, <a href='https://twitter.com/doctorcorces'>Ryan Corces</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


