---
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    theme: cerulean
params:
  ## will be override by render
  readin: "Sample.RDS"
  project_id: "Sample"
  pipe_dir: "/path/to/iSHARC"
  integr_dir: "/path/to/integration/folder"
---

This report is generated by the [iSHARC](https://github.com/yzeng-lol/iSHARC).
For more modularized downstream analyses, please refer to [COBE](https://www.pmcobe.ca/pipeline/618285d473c8c4003964393b).

```{r setup, include=FALSE}
##knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
project_id <- params$project_id
pipe_dir <- params$pipe_dir

## For RNA
rna_merged <- readRDS(paste0(params$integr_dir, "/rna/RNA_integrated_by_merging.RDS"))
rna_harmonized <- readRDS(paste0(params$integr_dir, "/rna/RNA_integrated_by_harmony.RDS"))
rna_anchored <- readRDS(paste0(params$integr_dir, "/rna/RNA_integrated_by_anchors.RDS"))

## For ATAC
atac_merged <- readRDS(paste0(params$integr_dir, "/atac/ATAC_integrated_by_merging.RDS"))
atac_harmonized <- readRDS(paste0(params$integr_dir, "/atac/ATAC_integrated_by_harmony.RDS"))
atac_anchored <-  readRDS(paste0(params$integr_dir, "/atac/ATAC_integrated_by_anchors.RDS"))

## For integrated RNA and ATAC
harmony_wnn <- readRDS(paste0(params$integr_dir, "/wnn/harmony/RNA_ATAC_integrated_by_WNN.RDS"))
anchor_wnn <- readRDS(paste0(params$integr_dir, "/wnn/anchor/RNA_ATAC_integrated_by_WNN.RDS"))

```

---
title: "`r project_id`_integrated_scMultiome_samples_QC_and_Primary_Results"
---

```{r include=FALSE}
## loading packages without showing
library(Seurat)
library(Signac)
library(dplyr)
library(ggplot2)
library(plotly)
library(knitr)
#library(SingleR)       ## auto annotation
#library(celldex)       ## annotation reference
library(heatmaply)
library(gapminder)
library(gplots)
```

## Number of cells
### Number of cells per sample
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

## number of cells per sample
dat <- rna_merged@meta.data %>%
       group_by(sample_id) %>%
       count(sample_id)

st <- paste0("Total number of cells: ", sum(dat$n))

p1 <- ggplot(dat, aes(y = n, x = sample_id, fill = sample_id)) + geom_bar(stat='identity')
p1 <- p1 + geom_text(aes(label = n), size = 5, vjust =  -0.5)
p1 <- p1 + labs(y = "Number of cells", x = "", subtitle = st)
p1 <- p1 + theme_classic() +  theme(legend.position = "none")

## count the cell cycle status for cells
dat <- rna_merged@meta.data %>%
       group_by(sample_id, ) %>%
       count(Phase)

p2 <- ggplot(dat, aes( y = n, x = sample_id, fill = Phase))  
p2 <- p2 + geom_bar(position="fill", stat="identity")
p2 <- p2 + labs(y = "Fraction of cells",  x = "Phase of cell cycle")
p2 <- p2 + guides(fill=guide_legend(title="Phase"))
p2 <- p2 + theme_classic() +  theme(legend.position = "right")

## combine figures for output
p1 + p2  & theme(plot.title = element_text(hjust = 0))
```

### Annotated cells per sample
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

## cancer vs no cancer cells
dat  <- rna_merged[[c("sample_id","copykat_anno")]]
dat_sum <- dat %>%
           group_by(sample_id, copykat_anno) %>%
           summarise(cnt = n())

p1 <- ggplot(dat_sum, aes( y = cnt, x = sample_id, fill = copykat_anno))  
p1 <- p1 + geom_bar(position="fill", stat="identity")
p1 <- p1 + labs(y = "Fraction of cells",  x = "")
p1 <- p1 + guides(fill=guide_legend(title="CopyKat prediction")) + theme_classic()

## cell annotation based single sample WNN clustering based prediction
dat  <- rna_merged[[c("sample_id","WNN_SingleR_anno")]]
dat_sum <- dat %>%
           group_by(sample_id, WNN_SingleR_anno) %>%
           summarise(cnt = n())

p2 <- ggplot(dat_sum, aes( y = cnt, x = sample_id, fill = WNN_SingleR_anno))  
p2 <- p2 + geom_bar(position="fill", stat="identity")
p2 <- p2 + labs(y = "Fraction of cells",  x = "")
p2 <- p2 + guides(fill=guide_legend(title="SingleR annotation")) + theme_classic()

## combine figures for output
p1 + p2  & theme(plot.title = element_text(hjust = 0.5))
```



## QC metrics across samples
### Selected Percentiles for QC metrics
* nCount_RNA: Number of transcripts detected per cell
* percent.mt: Percentage of reads originating from the mitochondrial genes
* nCount_ATAC: The number of unique nuclear fragments
* TSS.enrichment: The ratio of fragments centered at the TSS to fragments in TSS-flanking regions
* nucleosome_signal: Approximate ratio of mononucleosomal to nucleosome-free fragments

```{r echo = FALSE}
## display selected metrics
qc_df <- data.matrix(rna_merged[[c("nCount_RNA", "nCount_ATAC", "percent.mt", "TSS.enrichment", "nucleosome_signal")]])
probs_s <- c(0, 0.025, 0.25, 0.50, 0.75, 0.975, 1)
qc_percentiles <- apply(qc_df, 2, function(x) quantile(x, probs = probs_s, na.rm = TRUE))

#rownames(sum_adj) <- c("", "Q1", "Median", "Mean", "Q3", "Max")
ti <- paste("Number of call cells across samples: ", ncol(rna_merged))
kable(qc_percentiles, caption = ti)
```

### Violin plots for selected QC metrics
```{r  echo = FALSE, fig.width = 13, fig.height= 8}
VlnPlot(rna_merged, features = c("nCount_RNA", "nCount_ATAC", "percent.mt", "TSS.enrichment", "nucleosome_signal"), group.by = "sample_id", split.by = "sample_id", ncol = 3, log = TRUE, pt.size = 0)  & theme(axis.title.x = element_blank())
```

## RNA merge and integration
Integrating RNA by merging, Harmony and Seurat, respectively, and labeling by sample id (top) and clustering results (bottom).
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(rna_merged,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_merged")

p2 <- DimPlot(rna_harmonized,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_harmonized")
p3 <- DimPlot(rna_anchored,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))
```

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(rna_merged,  reduction = "umap", group.by = "RNA_integrated_snn_res.0.8",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_merged")

p2 <- DimPlot(rna_harmonized,  reduction = "umap", group.by = "RNA_integrated_snn_res.0.8",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_harmonized")

p3 <- DimPlot(rna_anchored,  reduction = "umap", group.by = "RNA_integrated_snn_res.0.8",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

rm(rna_merged, rna_harmonized, rna_anchored)
```

## ATAC merge and integration
Integrating ATAC by merging, Harmony and Seurat, respectively, and labeling by sample id (top) and clustering results (bottom).
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(atac_merged,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_merged")

p2 <- DimPlot(atac_harmonized,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_harmonized")
p3 <- DimPlot(atac_anchored,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))
```

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(atac_merged,  reduction = "umap", group.by = "ATAC_integrated_snn_res.0.8",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_merged")

p2 <- DimPlot(atac_harmonized,  reduction = "umap", group.by = "ATAC_integrated_snn_res.0.8",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_harmonized")

p3 <- DimPlot(atac_anchored,  reduction = "umap", group.by = "ATAC_integrated_snn_res.0.8",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

rm(atac_merged, atac_harmonized, atac_anchored)
```

## RNA and ATAC integration
### WNN based harmony/seurat integrated RNA and ATAC
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(harmony_wnn, reduction = "umap",  group.by = "RNA_integrated_snn_res.0.8", label = TRUE,         label.size = 2.5, repel = TRUE) + ggtitle("RNA_harmonized_integrated")
p2 <- DimPlot(harmony_wnn, reduction = "umap_atac", group.by = "ATAC_integrated_snn_res.0.8", label =          TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC_harmonized_integrated")
p3 <- DimPlot(harmony_wnn, reduction = "umap_wnn",  group.by = "RNA_ATAC_integrated_wsnn_res.0.8",            label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN_harmonized_RNA_ATAC")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

```

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

p1 <- DimPlot(anchor_wnn, reduction = "umap",  group.by = "RNA_integrated_snn_res.0.8", label = TRUE,         label.size = 2.5, repel = TRUE) + ggtitle("RNA_anchored_integrated")
p2 <- DimPlot(anchor_wnn, reduction = "umap_atac", group.by = "ATAC_integrated_snn_res.0.8", label =          TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC_anchored_integrated")
p3 <- DimPlot(anchor_wnn, reduction = "umap_wnn",  group.by = "RNA_ATAC_integrated_wsnn_res.0.8",            label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN_anchored_RNA_ATAC")
p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

```

### RNA weight for WNN based harmony/seurat integrated RNA and ATAC
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 7}
## morality weights
p1 <- VlnPlot(harmony_wnn, features = "RNA_integrated.weight", group.by = 'RNA_ATAC_integrated_wsnn_res.0.8', sort = F,         pt.size = 0.1)
p1 <- p1 + ggtitle("RNA weights for WNN harmonized RNA & ATAC") + labs(x = "WNN clusters") + NoLegend()

p2 <- VlnPlot(anchor_wnn, features = "RNA_integrated.weight", group.by = 'RNA_ATAC_integrated_wsnn_res.0.8', sort = F,          pt.size = 0.1)
p2 <- p2 + ggtitle("RNA weights for WNN anchored RNA & ATAC") + labs(x = "WNN clusters") + NoLegend()

p1 + p2  & theme(plot.title = element_text(hjust = 0.5))

```

###  Consistency of cell annotation per sample within each WNN clusters
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

### based on harmoney
dat  <- harmony_wnn[[c("RNA_ATAC_integrated_wsnn_res.0.8","WNN_SingleR_anno")]]
dat_sum <- dat %>%
           group_by(RNA_ATAC_integrated_wsnn_res.0.8, WNN_SingleR_anno) %>%
           summarise(cnt = n())

p1 <- ggplot(dat_sum, aes( y = cnt, x = RNA_ATAC_integrated_wsnn_res.0.8, fill = WNN_SingleR_anno))  
p1 <- p1 + geom_bar(position="fill", stat="identity")
p1 <- p1 + labs(y = "Fraction of cells",  title = "WNN harmonized RNA & ATAC", x = "WNN clusters")
p1 <- p1 + guides(fill=guide_legend(title="SingleR annotated per sample")) + theme_classic()

### based on anchors
dat  <- anchor_wnn[[c("RNA_ATAC_integrated_wsnn_res.0.8","WNN_SingleR_anno")]]
dat_sum <- dat %>%
           group_by(RNA_ATAC_integrated_wsnn_res.0.8, WNN_SingleR_anno) %>%
           summarise(cnt = n())

p2 <- ggplot(dat_sum, aes( y = cnt, x = RNA_ATAC_integrated_wsnn_res.0.8, fill = WNN_SingleR_anno))  
p2 <- p2 + geom_bar(position="fill", stat="identity")
p2 <- p2 + labs(y = "Fraction of cells",  title = "WNN anchored RNA & ATAC", x = "WNN clusters")
p2 <- p2 + guides(fill=guide_legend(title="SingleR annotated per sample")) + theme_classic()

p1 + p2  & theme(plot.title = element_text(hjust = 0.5))

```
