---
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    theme: cerulean
params:
  ## will be override by render
  integr_dir: "/path/to/integration/folder"
  pipe_dir: "/path/to/iSHARC"
---

This report was generated using [iSHARC](https://github.com/yzeng-lol/iSHARC).For additional downstream analyses and relevant tools, please refer to [COBE](https://www.pmcobe.ca/pipeline/618285d473c8c4003964393b).

```{r setup, include=FALSE}
##knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
pipe_dir <- params$pipe_dir

## For RNA
rna_merged <- readRDS(paste0(params$integr_dir, "/rna/RNA_integrated_by_merging.RDS"))
rna_harmonized <- readRDS(paste0(params$integr_dir, "/rna/RNA_integrated_by_harmony.RDS"))
rna_anchored <- readRDS(paste0(params$integr_dir, "/rna/RNA_integrated_by_anchors.RDS"))

## For ATAC
atac_merged <- readRDS(paste0(params$integr_dir, "/atac/ATAC_integrated_by_merging.RDS"))
atac_harmonized <- readRDS(paste0(params$integr_dir, "/atac/ATAC_integrated_by_harmony.RDS"))
atac_anchored <-  readRDS(paste0(params$integr_dir, "/atac/ATAC_integrated_by_anchors.RDS"))

## For integrated RNA and ATAC
harmony_wnn <- readRDS(paste0(params$integr_dir, "/wnn/harmony/RNA_ATAC_integrated_by_WNN.RDS"))
anchor_wnn <- readRDS(paste0(params$integr_dir, "/wnn/anchor/RNA_ATAC_integrated_by_WNN.RDS"))

```

---
title: "Integrated_samples_QC_and_Primary_Results"
---

```{r include=FALSE}
## loading packages without showing
library(Seurat)
library(Signac)
library(dplyr)
library(ggplot2)
library(plotly)
library(knitr)
#library(SingleR)       ## auto annotation
#library(celldex)       ## annotation reference
library(heatmaply)
#library(gapminder)
library(gplots)
```

## Cells Composition per Sample
### Number of cells and their cell cycle phase
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
#############################
## number of cells per sample
dat <- rna_merged@meta.data %>%
       group_by(sample_id) %>%
       count(sample_id)

st <- paste0("Total number of cells: ", sum(dat$n))

p1 <- ggplot(dat, aes(y = n, x = sample_id, fill = sample_id)) + geom_bar(stat='identity')
p1 <- p1 + geom_text(aes(label = n), size = 5, vjust =  -0.5)
p1 <- p1 + labs(y = "Number of cells", x = "", subtitle = st)
p1 <- p1 + theme_classic() +  theme(legend.position = "none")

#######################################
## count the cell cycle status for cells
dat <- rna_merged@meta.data %>%
       group_by(sample_id, ) %>%
       count(Phase)

p2 <- ggplot(dat, aes( y = n, x = sample_id, fill = Phase))  
p2 <- p2 + geom_bar(position="fill", stat="identity")
p2 <- p2 + labs(y = "Fraction of cells",  x = "Phase of cell cycle")
p2 <- p2 + guides(fill=guide_legend(title="Phase"))
p2 <- p2 + theme_classic() +  theme(legend.position = "right")

## combine figures for output
p1 + p2  & theme(plot.title = element_text(hjust = 0))
```

### Cells annotation
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

## cancer vs no cancer cells
dat  <- rna_merged[[c("sample_id","WNN_clusters_copykat_annot")]]
dat_sum <- dat %>%
           group_by(sample_id, WNN_clusters_copykat_annot) %>%
           summarise(cnt = n())

p1 <- ggplot(dat_sum, aes( y = cnt, x = sample_id, fill = WNN_clusters_copykat_annot))  
p1 <- p1 + geom_bar(position="fill", stat="identity")
p1 <- p1 + labs(y = "Fraction of cells",  x = "")
p1 <- p1 + guides(fill=guide_legend(title="CopyKat prediction")) + theme_classic()

## cell annotation based single sample WNN clustering based prediction
dat  <- rna_merged[[c("sample_id","WNN_clusters_singler_annot")]]
dat_sum <- dat %>%
           group_by(sample_id, WNN_clusters_singler_annot) %>%
           summarise(cnt = n())

p2 <- ggplot(dat_sum, aes( y = cnt, x = sample_id, fill = WNN_clusters_singler_annot))  
p2 <- p2 + geom_bar(position="fill", stat="identity")
p2 <- p2 + labs(y = "Fraction of cells",  x = "")
p2 <- p2 + guides(fill=guide_legend(title="SingleR annotation")) + theme_classic()

## combine figures for output
p1 + p2  & theme(plot.title = element_text(hjust = 0.5))
```

## QC metrics across samples
### Selected percentiles for QC metrics
* nCount_RNA: The number of transcripts detected per cell.
* pct_MT: The percentage of reads originating from the mitochondrial genes.
* nCount_ATAC: The number of unique nuclear fragments.
* TSS_Enrichment: The ratio of fragments centered at TSS to those in TSS-flanking regions.
* Nucleosome_Signal: The approximate ratio of mononucleosomal to nucleosome-free fragments.

```{r echo=FALSE, message = FALSE, warning = FALSE}
## display selected metrics
qc_df <- data.matrix(rna_merged[[c("nCount_RNA", "pct_MT", "nCount_ATAC", "TSS_Enrichment", "Nucleosome_Signal")]])
probs_s <- c(0, 0.025, 0.25, 0.50, 0.75, 0.975, 1)
qc_percentiles <- apply(qc_df, 2, function(x) quantile(x, probs = probs_s, na.rm = TRUE))

#rownames(sum_adj) <- c("", "Q1", "Median", "Mean", "Q3", "Max")
ti <- paste("Total Number of cells across samples: ", ncol(rna_merged))
kable(qc_percentiles, caption = ti)

```

### Violin plots for selected QC metrics
```{r  echo = FALSE, message = FALSE, warning = FALSE, fig.width = 13, fig.height= 8}
VlnPlot(rna_merged, features = c("nCount_RNA", "pct_MT", "nCount_ATAC", "TSS_Enrichment", "Nucleosome_Signal"), group.by = "sample_id", split.by = "sample_id", ncol = 3, log = TRUE, pt.size = 0)  & theme(axis.title.x = element_blank())
```

## Horizontally Integrated RNA
RNA integration across samples is performed using three approaches: simple merging, Harmony, and Seurat integration.The results are visualized below with cells labeled by sample ID (top panels) and clustering outcomes (bottom panels)


```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(rna_merged,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + NoLegend() +  ggtitle("RNA_merged")

p2 <- DimPlot(rna_harmonized,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + NoLegend() + ggtitle("RNA_harmonized")
p3 <- DimPlot(rna_anchored,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))
```

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(rna_merged,  reduction = "umap", group.by = "RNA_integrated_clusters",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_merged")

p2 <- DimPlot(rna_harmonized,  reduction = "umap", group.by = "RNA_integrated_clusters",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_harmonized")

p3 <- DimPlot(rna_anchored,  reduction = "umap", group.by = "RNA_integrated_clusters",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("RNA_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

rm(rna_merged, rna_harmonized, rna_anchored)
```

## Horizontally Integrated ATAC
ATAC integration across samples is performed using three approaches: simple merging, Harmony, and Seurat integration.The results are visualized below with cells labeled by sample ID (top panels) and clustering outcomes (bottom panels)

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(atac_merged,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + NoLegend() + ggtitle("ATAC_merged")

p2 <- DimPlot(atac_harmonized,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE) + NoLegend()  + ggtitle("ATAC_harmonized")
p3 <- DimPlot(atac_anchored,  reduction = "umap", group.by = "sample_id",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))
```

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(atac_merged,  reduction = "umap", group.by = "ATAC_integrated_clusters",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_merged")

p2 <- DimPlot(atac_harmonized,  reduction = "umap", group.by = "ATAC_integrated_clusters",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_harmonized")

p3 <- DimPlot(atac_anchored,  reduction = "umap", group.by = "ATAC_integrated_clusters",
                label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("ATAC_anchored")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

rm(atac_merged, atac_harmonized, atac_anchored)
```

## vertically Integrated ATAC and RNA
### WNN on top of harmony (top) and Seurat (bottom) integrated RNA and ATAC
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
p2 <- DimPlot(harmony_wnn, reduction = "umap",  group.by = "RNA_integrated_clusters", label = TRUE,         label.size = 2.5, repel = TRUE) + ggtitle("RNA_harmonized")
p1 <- DimPlot(harmony_wnn, reduction = "umap_atac", group.by = "ATAC_integrated_clusters", label =          TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC_harmonized")
p3 <- DimPlot(harmony_wnn, reduction = "umap_wnn",  group.by = "multiple_samples_WNN_clusters",            label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN_harmonized_RNA_ATAC")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

```

```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

p2 <- DimPlot(anchor_wnn, reduction = "umap",  group.by = "RNA_integrated_clusters", label = TRUE,         label.size = 2.5, repel = TRUE) + ggtitle("RNA_anchored")
p1 <- DimPlot(anchor_wnn, reduction = "umap_atac", group.by = "ATAC_integrated_clusters", label =          TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC_anchored")
p3 <- DimPlot(anchor_wnn, reduction = "umap_wnn",  group.by = "multiple_samples_WNN_clusters",            label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN_anchored_RNA_ATAC")
p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

```

### RNA weights for WNN harmonized (top) and anchored (bottom) RNA and ATAC
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 7}
## morality weights
p1 <- VlnPlot(harmony_wnn, features = "RNA_integrated.weight", group.by = 'multiple_samples_WNN_clusters', sort = F,         pt.size = 0.1)
p1 <- p1 + ggtitle("RNA weights for WNN harmonized RNA & ATAC") + labs(x = "WNN clusters") + NoLegend()

p2 <- VlnPlot(anchor_wnn, features = "RNA_integrated.weight", group.by = 'multiple_samples_WNN_clusters', sort = F,          pt.size = 0.1)
p2 <- p2 + ggtitle("RNA weights for WNN anchored RNA & ATAC") + labs(x = "WNN clusters") + NoLegend()

p1 + p2  & theme(plot.title = element_text(hjust = 0.5))

```

### Cell annotation distribution within each WNN-integrated cluster
Note: Cell annotations were assigned based on individual samples before integration of multiple samples.
```{r  echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}

### based on harmoney
dat  <- harmony_wnn[[c("multiple_samples_WNN_clusters", "WNN_clusters_singler_annot")]]
dat_sum <- dat %>%
           group_by(multiple_samples_WNN_clusters,  WNN_clusters_singler_annot) %>%
           summarise(cnt = n())

p1 <- ggplot(dat_sum, aes( y = cnt, x = multiple_samples_WNN_clusters, fill = WNN_clusters_singler_annot))  
p1 <- p1 + geom_bar(position="fill", stat="identity")
p1 <- p1 + labs(y = "Fraction of cells",  title = "WNN harmonized RNA & ATAC", x = "WNN clusters")
p1 <- p1 + guides(fill=guide_legend(title="SingleR annotated per sample")) + theme_classic()

### based on anchors
dat  <- anchor_wnn[[c("multiple_samples_WNN_clusters","WNN_clusters_singler_annot")]]
dat_sum <- dat %>%
           group_by(multiple_samples_WNN_clusters, WNN_clusters_singler_annot) %>%
           summarise(cnt = n())

p2 <- ggplot(dat_sum, aes( y = cnt, x = multiple_samples_WNN_clusters, fill = WNN_clusters_singler_annot))  
p2 <- p2 + geom_bar(position="fill", stat="identity")
p2 <- p2 + labs(y = "Fraction of cells",  title = "WNN anchored RNA & ATAC", x = "WNN clusters")
p2 <- p2 + guides(fill=guide_legend(title="SingleR annotated per sample")) + theme_classic()

p1 + p2  & theme(plot.title = element_text(hjust = 0.5))

```

## List of main integrated outputs
Note: additional figures are available in the directory of : workdir/integrated_samples/!!

### Differently integrated Seurat objects
* rna/RNA_integrated_by_[merging | anchors | harmony].RDS: These are Seurat objects of horizontally integrated RNA across multiple samples by applying merge, Seurat, Harmony, respectively.  
* atac/ATAC_integrated_by_[merging | anchors | harmony].RDS: These are Seurat objects of horizontally integrated ATAC across multiple samples by applying merge, Seurat, Harmony, respectively.
* wnn/harmony/RNA_ATAC_integrated_by_WNN.RDS: This Seurat object includes the vertically integrated RNA and ATAC across multiple sample after horizontally integrate each modality using Harmony.
* wnn/achor/RNA_ATAC_integrated_by_WNN.RDS: This Seurat object includes the vertically integrated RNA and ATAC across multiple sample after horizontally integrate each modality using Seurat.

### RMD file
* Integrated_samples_QC_and_Primary_Results.Rmd: The Rmd file for generating this HTML report.
