---
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    theme: cerulean
params:
  ## will be override by render
  readin: "Sample.RDS"
  sample_id: "Sample"
  pipe_dir: "/path/to/iSHARC"
---

```{r setup, include=FALSE}
##knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
## read in scMultiome.RDS
scMultiome <- readRDS(params$readin)
sample_id <- params$sample_id
pipe_dir <- params$pipe_dir
```

This report was generated using [iSHARC](https://github.com/yzeng-lol/iSHARC).For additional downstream analyses and relevant tools, please refer to [COBE](https://www.pmcobe.ca/pipeline/618285d473c8c4003964393b).

---
title: "`r sample_id`_scMultiome_QC_and_Primary_Results"
---

```{r include=FALSE}
## loading packages without showing
library(Seurat)
library(Signac)
library(dplyr)
library(ggplot2)
library(plotly)
library(knitr)
library(SingleR)       ## auto annotation
library(celldex)       ## annotation reference
library(heatmaply)
#library(gapminder)
library(gplots)
library(ggpubr)
library(ggraph)
library(tidygraph)
library(clusterProfiler)

if(!require("KEGG.db"))  install.packages(paste0(pipe_dir, "/workflow/dependencies/KEGG.db_1.0.tar.gz"))
library(KEGG.db)

```

## QC metrics and filtering
### Selected percentiles for QC metrics
* nCount_RNA: The number of transcripts detected per cell.
* pct_MT: The percentage of reads originating from the mitochondrial genes.
* nCount_ATAC: The number of unique nuclear fragments.
* TSS_Enrichment: The ratio of fragments centered at TSS to those in TSS-flanking regions.
* Nucleosome_Signal: The approximate ratio of mononucleosomal to nucleosome-free fragments.

```{r echo = FALSE}
## display selected metrics

qc_df <- data.matrix(scMultiome[[c("nCount_RNA", "pct_MT", "nCount_ATAC", "TSS_Enrichment", "Nucleosome_Signal")]])
probs_s <- c(0, 0.025, 0.25, 0.50, 0.75, 0.975, 1)
qc_percentiles <- apply(qc_df, 2, function(x) quantile(x, probs = probs_s, na.rm = TRUE))

#rownames(sum_adj) <- c("", "Q1", "Median", "Mean", "Q3", "Max")
ti <- paste("Number of joint called cells by Cellranger-ARC: ", ncol(scMultiome))
kable(qc_percentiles, caption = ti)
```

### Violin plots for selected QC metrics
```{r  echo = FALSE, fig.width = 13}
VlnPlot(scMultiome, features = c("nCount_RNA", "pct_MT", "nCount_ATAC", "TSS_Enrichment", "Nucleosome_Signal"),       ncol = 5, log = TRUE, pt.size = 0, group.by = "orig.ident") + NoLegend() & theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

### Second-round QC filtering assessment
The automatic QC metric cutoffs are determined based on the suggested thresholds and the corresponding median Â± 3*MAD (Median Absolute Deviation), as outlined below. The subsequent results for the cells in this report depend on whether the second-round QC filtering is enabled in the configuration file.

* nCount_RNA:  [median - 3*MAD, median + 3*MAD]
* nCount_ATAC: [median - 3*MAD, median + 3*MAD]
* pct_MT:  < min(20, median + 3*MAD)
* TSS_Enrichment: > max(1, median - 3*MAD)
* Nucleosome_Signal: < min(2, median + 3*MAD)

```{r echo = FALSE}

## display selected metrics
kable(scMultiome@misc[["second_QC_assessment"]])

```

## Cell cycle assessment
* Exercise caution when analyzing cells undergoing differentiation processes (e.g., [hematopoiesis](https://satijalab.org/seurat/articles/cell_cycle_vignette.html#alternate-workflow)).
* PCA plots are provided both before and after regressing out the effects of the cell cycle. Only cell cycle-related genes are used for the PCA analysis.
```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
## bar plot for the number of cells per cell cycle status
if(TRUE){
cc_cnt <- table(scMultiome$Phase)
cc_pct <- paste(round(100 * cc_cnt/sum(cc_cnt), 2), "%", sep="")
cc_group <- factor(names(cc_cnt), levels = c("G1", "S", "G2M", "Undecided"))

dat <- data.frame(cc_cnt, cc_pct, cc_group)

p1 <- ggplot(dat, aes(y = cc_cnt, x = cc_group, fill = cc_group)) + geom_bar(stat='identity')
p1 <- p1 + geom_text(aes(label = cc_cnt), size = 5, vjust =  1.5)
p1 <- p1 + geom_text(aes(label = cc_pct), size = 5, vjust = -0.2)
p1 <- p1 + labs(y = "Number of cells", x = "Cell cycle phases")
p1 <- p1 + theme_classic() +   theme(legend.position = "none")

## PCA plots before and after regressing out the cell cycle effects
## cell cycle genes
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

## without regressing out cell cycle
cc_ori <- SCTransform(
                scMultiome,
                assay = 'RNA',
                new.assay.name = 'SCT_cc',
                vars.to.regress = c('pct_MT'),
                vst.flavor = "v2", verbose = FALSE
                )

cc_ori <- RunPCA(cc_ori, features = c(s.genes, g2m.genes))
p2 <- DimPlot(cc_ori, group.by = "Phase") + ggtitle("Before")

## regressing out cell cycle
DefaultAssay(scMultiome) <- "SCT"   ## which has regressed out cell cycle by default
cc_reg <- RunPCA(scMultiome, features = c(s.genes, g2m.genes))
p3 <- DimPlot(cc_reg, reduction = "pca", group.by = "Phase") + ggtitle("After")

## combine figures for output
p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))
}
```

## Integration and Clustering
### ATAC, RNA and integrated WNN clustering
```{r, echo=FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(scMultiome, reduction = "umap.atac", group.by = "ATAC_clusters", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC")

p2 <- DimPlot(scMultiome, reduction = "umap.rna",  group.by = "RNA_clusters", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("RNA")

p3 <- DimPlot(scMultiome, reduction = "umap.wnn", group.by = "WNN_clusters",  label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

```

### Modality weights for integrated WNN clusters
Higher ATAC weights indicate that the epigenome has a greater influence on the corresponding cell clusters. Similarly, higher RNA weights suggest that the transcriptome plays a more significant role in the corresponding cell clusters.

```{r, echo=FALSE, fig.width = 15, fig.height= 3}
VlnPlot(scMultiome, features = "ATAC.weight", group.by = 'WNN_clusters', sort = FALSE, pt.size = 0.1) +  ggtitle("ATAC weights") & theme(axis.text.x = element_blank(), axis.title.x = element_blank())

VlnPlot(scMultiome, features = "SCT.weight", group.by = 'WNN_clusters', sort = FALSE, pt.size = 0.1) + ggtitle("RNA weights")

```

### Clustering changes across individual and integrated modalities
The interactive Sankey plot below enables the tracking of clustering changes between individual and integrated modalities.
```{r, echo=FALSE, fig.width = 8.5, fig.height= 5}
# library(plotly)
Cell_ID <- rownames(scMultiome@meta.data)
RNA_Clusters <- paste0("RNA_", scMultiome@meta.data$RNA_clusters)
ATAC_Clusters <- paste0("ATAC_", scMultiome@meta.data$ATAC_clusters)
WNN_Clusters <- paste0("WNN_", scMultiome@meta.data$WNN_clusters)

df <- data.frame(Cell_ID, RNA_Clusters, ATAC_Clusters, WNN_Clusters)
df <- as_tibble(df)

## unique nodes and indexing
nodes <- df %>%
         pivot_longer(-Cell_ID, values_to = "name_node") %>%
         distinct(name_node) %>%
         arrange(name_node) %>%
         mutate(idx = (1:n()) - 1)

## links with source, target and values (count)
links <- bind_rows(df %>% dplyr::select(source = ATAC_Clusters, target = RNA_Clusters),
                   df %>% dplyr::select(source = RNA_Clusters, target = WNN_Clusters)) %>%
         group_by(source, target) %>% dplyr::count(target) %>%
         mutate(value = n) %>% dplyr::select(!n) %>%  ungroup()

links$source <- nodes$idx[match(links$source, nodes$name_node)]
links$target <- nodes$idx[match(links$target, nodes$name_node)]

g <- plot_ly(
     type = "sankey",
     orientation = "h",
     node = list(label = nodes$name_node, pad = 15, thickness = 15),
     link = as.list(links))
g
```

## Auto cell annotation
### Annotation using publicly available references
The Blueprint/ENCODE reference from the R package [celldex](https://bioconductor.org/packages/3.17/data/experiment/vignettes/celldex/inst/doc/userguide.html) is used for automatically annotating each cell cluster using [SingleR](https://github.com/dviraran/SingleR).

```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 6}
## just in case the scMultiome.RDS didn't automatic annotated
if(is.na(match("WNN_clusters_singler_annot", colnames(scMultiome@meta.data)))){

## pre-downloaded data has been added to ./iSHARC/workflow/dependencies/
ref_rds <- paste0(pipe_dir, "/workflow/dependencies/BlueprintEncodeData.RDS")
anno_ref <- readRDS(ref_rds)

## fetch SCT normalized GEX matrix
expr <- GetAssayData(object = scMultiome, assay = "SCT", slot = "data")

### using ENCODE
expr_anno <- SingleR(test = expr, ref = anno_ref, labels = anno_ref$label.main, clusters =  Idents(scMultiome))

## match cluster labels and annotated labels
idx_m <- match(Idents(scMultiome), rownames(expr_anno))

## add labels scMultiome object
scMultiome[["WNN_clusters_singler_annot"]] <- expr_anno$labels[idx_m]
}

## umap
p1 <- DimPlot(scMultiome, reduction = "umap.wnn", group.by = "WNN_clusters_singler_annot",  label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("")

## bar plot with WNN labels and singleR labels
dat  <- scMultiome[[c("WNN_clusters", "WNN_clusters_singler_annot")]]
dat_sum <- dat %>%
           group_by(WNN_clusters, WNN_clusters_singler_annot) %>%
           summarise(cnt = n())

p2 <- ggplot(dat_sum, aes(fill = WNN_clusters, y = cnt, x = WNN_clusters_singler_annot))  
p2 <- p2 + geom_bar(position="stack", stat="identity")
p2 <- p2 + labs(y = "Number of cells", x = "SingleR annotated cell types")
p2 <- p2 + guides(fill=guide_legend(title="WNN clusters")) + theme_classic()

p1 + p2   & theme(plot.title = element_text(hjust = 0.5))
```

### Inferring tumor and normal cells
[Copykat](https://github.com/navinlabcode/copykat) is employed to classify cells into the following categories: aneuploid (tumor cells); diploid (stromal normal cells); not.defined (Cells taht cannot be to be predicted by copykat); not.predicted (cells excluded from copykat prediction).

```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 6}
if(TRUE){
## umap
p1 <- DimPlot(scMultiome, reduction = "umap.wnn", group.by = "WNN_clusters_copykat_annot",  label = TRUE, label.size = 2.5, repel = TRUE)  + ggtitle("")

## bar plot with WNN labels and singleR labels
dat  <- scMultiome[[c("WNN_clusters", "WNN_clusters_copykat_annot")]]
dat_sum <- dat %>%
           group_by(WNN_clusters, WNN_clusters_copykat_annot) %>%
           summarise(cnt = n())

p2 <- ggplot(dat_sum, aes(fill = WNN_clusters, y = cnt, x = WNN_clusters_copykat_annot))  
p2 <- p2 + geom_bar(position="stack", stat="identity")
p2 <- p2 + labs(y = "Number of cells", x = "Copykat predict cell types")
p2 <- p2 + guides(fill=guide_legend(title="WNN clusters")) + theme_classic()

p1 + p2   & theme(plot.title = element_text(hjust = 0.5))
}
```


## Cluster-specific genes
### Top 5 upregualted genes per WNN cluster
The ATAC peaks linked to these top 5 genes can be found in the file `r sample_id`_top5_DEGs_linked_peaks.csv. This file will be empty if no genes passed the preset cut-offs.

```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 6}

if(length(scMultiome@assays$SCT@misc$DEGs_top5) > 0){
source(paste0(pipe_dir, "/workflow/scripts/DoMutiBarHeatmap.R"))   ## function for multiple bars

DoMultiBarHeatmap(scMultiome, features = scMultiome@assays$SCT@misc$DEGs_top5, assay = 'SCT',
                  group.by='WNN_clusters_singler_annot', label = FALSE,
                  additional.group.by = c('WNN_clusters_copykat_annot', "Phase", 'ATAC_clusters',  'RNA_clusters', 'WNN_clusters'),
                  additional.group.sort.by = c('WNN_clusters'))
}

```

### Functional enrichment analysis for WNN cluster-specific genes
```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 12}

if(TRUE){
    ## ID convert function
    id_convert <- function(x){
      library(clusterProfiler)
      g_symbol <- rownames(x)
      if(length(g_symbol) <= 10){
        ## requiring at least 10 genes
        g_ezid <- NULL
      } else {
      g_con <- bitr(g_symbol, fromType = "SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")
      g_ezid <- g_con$ENTREZID
      }

      return(g_ezid)
    }

    deg_list <- lapply(scMultiome@assays$SCT@misc$DEGs, id_convert)

    ## KEGG enrichment
    compKEGG <- compareCluster(geneCluster   = deg_list,
                               fun           = "enrichKEGG",
                               pvalueCutoff  = 0.05,
                               pAdjustMethod = "BH",
                               use_internal_data =T)    ## using local build library "KEGG.db"

    p1 <- dotplot(compKEGG, showCategory = 2, title = "KEGG pathway enrichment") + labs(x = "") + scale_x_discrete(guide = guide_axis(angle = 45))


     ## GO enrichment
    compGO <- compareCluster(geneCluster   = deg_list ,
                             fun           = "enrichGO",
                             OrgDb='org.Hs.eg.db',
                             pvalueCutoff  = 0.05,
                             pAdjustMethod = "BH")

    p2 <- dotplot(compGO, showCategory = 1, title = "GO enrichment ") + labs(x = "") + scale_x_discrete(guide = guide_axis(angle = 45))

    ## combine p1 and p2 by two rows
    p1 / p2
}

```


## Cluster-specific regulatory regions
Top 5 enriched TF motifs per WNN cluster.
* NOTE: The heatmap is generated from the combined top-enriched TF motifs, saved in Seurat_object\@assays\$ATAC\@misc\$DARs_motif_hm. Values with -log10(p_adj) > 10 are capped at 10. The heatmap will be empty if no motifs are enriched or if they do not pass the preset cutoffs.

```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 9.25, fig.height= 4.5}
{
mat <- scMultiome@assays$ATAC@misc$DARs_motif_hm

if (isTRUE(nrow(mat) > 0)){
mat[mat > 10] <- 10

heatmaply(mat,
        dendrogram = "column",
        xlab = "Enriched Motifs", ylab = "",
        main = "",
        scale = "none",
        grid_color = "white",
        grid_width = 0.00001,
        titleX = FALSE,
        hide_colorbar = TRUE,
        branches_lwd = 0.1,
        label_names = c("Cluster:", "Motif:", "-log10(p_adj):"),
        fontsize_row = 10, fontsize_col = 10,
        labCol = colnames(mat),
        labRow = rownames(mat),
        #heatmap_layers = theme(axis.line=element_blank()),
        ## costomize color
        scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
          low = "white",
          high = "red",
          #midpoint = 5,
          limits = c(0, max(max(mat)))
          )
       )
     }

}
```


## Gene Regulatory Network
The [Pando](https://quadbio.github.io/Pando/) is employed for GRN analysis:

* Edges Color: Darkgrey (Inhibitory); Orange (Activating)
* Nodes Color: Lightgrey (DEG); Brown (TF); Brown with Black circle (DEG & TF)
* Nodes Size : Based on their centrality in the graph
* It will be empty if no GRNs are identified with preset cutoffs.

### GRN based on combined WNN cluster-specific genes
```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 7, fig.height= 7}
if(TRUE){
## GRN
edge_color = c('-1'='darkgrey', '1'='orange')  ## darkgrey for repression and orange for activate
node_fill_color = c("0" = "lightgrey", "1" = "Brown")
node_color = c("0" = "white", "1" = "Black")

if(length(scMultiome@misc$Combined_DEGs_GRN) > 0){
grn <- ggraph(scMultiome@misc$Combined_DEGs_GRN[[1]], layout='fr')
grn <- grn + geom_edge_diagonal(aes(color=factor(dir)), width = 0.2) + scale_edge_color_manual(values=edge_color)
grn <- grn + geom_node_point(aes(size = centrality, fill = factor(isTF), color = factor(isTargetTF)), shape=21)
grn <- grn + scale_fill_manual(values = node_fill_color) + scale_color_manual(values = node_color)
grn <- grn + geom_node_text(aes(label = name_label, size = centrality), repel=T)
grn <- grn + theme_void() + theme(legend.position = "none") # + ggtitle("Combined_DEGs_GRN")

## to avoid mismatching with previous "g"
grn

}

}
```

### GRN based on individual WNN cluster-specific genes
```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 7, fig.height = 15}
if(TRUE){
## GRN
edge_color = c('-1'='darkgrey', '1'='orange')  ## darkgrey for repression and orange for activate
node_fill_color = c("0" = "lightgrey", "1" = "Brown")
node_color = c("0" = "white", "1" = "Black")

L <- length(scMultiome@misc$Cluster_DEGs_GRN)
if(L > 0){

g_list <- list()
k = 1
for(i in 1:L)
  {

  g <- ggraph(scMultiome@misc$Cluster_DEGs_GRN[[i]], layout='fr')
  g <- g + geom_edge_diagonal(aes(color=factor(dir)), width = 0.2) +  scale_edge_color_manual(values=edge_color)
  g <- g + geom_node_point(aes(size = centrality, fill = factor(isTF), color = factor(isTargetTF)), shape=21)
  g <- g + scale_fill_manual(values = node_fill_color) + scale_color_manual(values = node_color)
  g <- g + geom_node_text(aes(label = name), repel=T)
  g <- g + theme_void() + theme(legend.position = "none")  + ggtitle(names(scMultiome@misc$Cluster_DEGs_GRN)[i])

  g_list[[k]] <- g

   ## to add space between plots
  g_void <- ggplot() + theme_void()
  g_list[[k + 1]] <-  g_void  
  k <- k + 2
  }
print(ggarrange(plotlist = g_list, ncol = 1, heights = rep(c(1, 0.1), L)))
}
}
```

## List of main outputs
Note: additional tables and figures are available in the directory of individual samples: workdir/individual_samples/`r sample_id`!!

### Three depth of Seurat objects
* `r sample_id`_initial_seurat_object.RDS: The initial Seurat object containing RNA and ATAC assays, with peaks re-called using MACS. No additional analyses are performed.
* `r sample_id`_vertically_integrated_seurat_object.RDS: This Seurat object includes the optional second-round QC, cell cycle correction, normalization, and clustering for RNA and ATAC data. RNA and ATAC data are integrated using WNN.
* `r sample_id`_extended_seurat_object.RDS : The extended Seurat object contains the results for additional assays based on the WNN integrated clusters, including intermediate and final results for all assays, particularly:
  - Seurat_object\@assays$ATAC: The ATAC assays performed by calling peaks with MACS.
  - Seurat_object\@assays$SCT\@misc\$DEGs: The WNN cluster-specific differentially expressed genes (DEGs)
  - Seurat_object\@assays$ATAC\@misc\$DARs: The WNN cluster-specific differential accessible regions (DARs)
  - Seurat_object\@assays$ATAC\@misc\$DARs_motif: The motif enrichment results for top cluster-specific DARs (p_adj < 0.005 )
  - Seurat_object\@assays$ATAC\@links: The peaks linked to the top 5 WNN cluster-specific DEGs
  - Seurat_object\@assays$ATAC\@links: The peaks linked to the top 5 WNN cluster-specific DEGs
  - scMultiome\@misc\$Combined_DEGs_GRN: tbl_graph object for combined DEGs
  - scMultiome\@misc\$Cluster_DEGs_GRN:  tbl_graph objects for WNN cluster-specific DEGs

### Tables
* `r sample_id`_extended_metaData.csv : A metadata table that includes QC metrics, cell cycle phases, clustering results for RNA, ATAC, and WNN integration, SingleR annotations, and CopyKAT tumor cell predictions for each cell.
* `r sample_id`*DEGs_GRN_Modules.csv : CSV files for the DEGs-based GRN modules.

### RMD file
* `r sample_id`_QC_and_Primary_Results.Rmd: The Rmd file for generating this HTML report.
