---
#title: "scMultiome_QC_Report_per_Sample"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    theme: cerulean
params:
  ## will be override by render
  readin: "Sample.RDS"
  sample_id: "Sample"
---

This report is generated by the [iSHARC](https://github.com/yzeng-lol/iSHARC)

```{r setup, include=FALSE}
##knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
## read in scMultiome.RDS
scMultiome <- readRDS(params$readin)
sample_id <- params$sample_id
```

---
title: "`r sample_id`_scMultiome_QC_Report"
---

```{r include=FALSE}
## loading packages without showing
library(Seurat)
library(Signac)
library(dplyr)
library(ggplot2)
library(plotly)
library(knitr)
#library(heatmaply)
#library(gapminder)
#library(gplots)
```

## QC metrics
### Selected Quantiles for QC metrics
* nCount_RNA: Number of transcripts detected per cell
* percent.mt: Percentage of reads originating from the mitochondrial genes
* nCount_ATAC: The number of unique nuclear fragments
* TSS.enrichment: The ratio of fragments centered at the TSS to fragments in TSS-flanking regions
* nucleosome_signal: Approximate ratio of mononucleosomal to nucleosome-free fragments

```{r echo = FALSE}
## display selected metrics
library(knitr)
qc_df <- data.matrix(scMultiome [[c("nCount_RNA","percent.mt", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal")]])
qc_quantiles <- apply(qc_df, 2, function(x) quantile(x, probs = seq(0, 1, 0.05), na.rm = TRUE))

## select rows for displaying
qc_quantiles <- qc_quantiles[c(1, 2, 6, 11, 16, 20 , 21), ]

#rownames(sum_adj) <- c("", "Q1", "Median", "Mean", "Q3", "Max")
ti <- paste("Number of joint called cells by Cellranger-ARC: ", ncol(scMultiome))
kable(qc_quantiles, caption = ti)
```

### Violin plots for selected QC metrics
```{r  echo = FALSE, fig.width = 13}
VlnPlot(scMultiome, features = c("nCount_RNA","percent.mt", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"), ncol = 5, log = TRUE, pt.size = 0, group.by = "orig.ident") + NoLegend()
```

## Cellcycle assessment
PCA plots before and after regressing out the effects of cell cycle, only cell cycle genes are used for PCA analysis.
```{r, echo=FALSE, message = FALSE, warning = FALSE, fig.width = 15, fig.height= 5}
## cell cycel genes
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

## without regressing out cell cycle 
cc_ori <- SCTransform(
    scMultiome,
    assay = 'RNA',
    new.assay.name = 'SCT_cc',
    vars.to.regress = c('percent.mt'), #, 'nFeature_RNA', 'nCount_RNA')
    verbose = F    ## doesn't print out the progress bars
    )
cc_ori <- RunPCA(cc_ori, features = c(s.genes, g2m.genes))
p1 <- DimPlot(cc_ori, group.by = "Phase") + ggtitle("Before")

## regressing out cell cycle 
DefaultAssay(scMultiome) <- "SCT"   ## which has regressed out cell cycle by default 
cc_reg <- RunPCA(scMultiome, features = c(s.genes, g2m.genes))
p2 <- DimPlot(cc_reg, reduction = "pca", group.by = "Phase") + ggtitle("After")

## combine figures for output
p1 + p2   & theme(plot.title = element_text(hjust = 0.5))

```

## Clustering results 
### Based on RNA, ATAC alone and WNN integreated 
```{r, echo=FALSE, fig.width = 15, fig.height= 5}
p1 <- DimPlot(scMultiome, reduction = "umap.rna",  group.by = "SCT_snn_res.0.8", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("RNA")
p2 <- DimPlot(scMultiome, reduction = "umap.atac", group.by = "ATAC_snn_res.0.8", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("ATAC")
p3 <- DimPlot(scMultiome, reduction = "wnn.umap", group.by = "wsnn_res.0.8",  label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN")

p1 + p2 + p3  & theme(plot.title = element_text(hjust = 0.5))

```

### Modality weights for integreated clustering
```{r, echo=FALSE, fig.width = 15, fig.height= 3}
VlnPlot(scMultiome, features = "ATAC.weight", group.by = 'wsnn_res.0.8', sort = FALSE, pt.size = 0.1) +  ggtitle("ATAC weights") 

VlnPlot(scMultiome, features = "SCT.weight", group.by = 'wsnn_res.0.8', sort = FALSE, pt.size = 0.1) + ggtitle("RNA weights") 
 
```
