---
#title: "scMultiome_QC_Report_per_Sample"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    theme: cerulean
params:
  ## will be override by render
  readin: "Lung.RDS"
  sample_id: "Lung"
---

This report is generated by the [iSHARC](https://github.com/yzeng-lol/iSHARC)

```{r setup, include=FALSE}
##knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
## read in scMultiome.RDS
scMultiome <- readRDS(params$readin)
sample_id <- params$sample_id
```

---
title: "`r sample_id`_scMultiome_QC_Report"
---

```{r include=FALSE}
## loading packages without showing
library(Seurat)
library(Signac)
library(dplyr)
library(ggplot2)
library(plotly)
library(knitr)
#library(heatmaply)
#library(gapminder)
#library(gplots)
```

## QC metrics
### Selected Quantiles for QC metrics
* nCount_RNA: Number of transcripts detected per cell
* percent.mt: Percentage of reads originating from the mitochondrial genes
* nCount_ATAC: The number of unique nuclear fragments
* TSS.enrichment: The ratio of fragments centered at the TSS to fragments in TSS-flanking regions
* nucleosome_signal: Approximate ratio of mononucleosomal to nucleosome-free fragments

```{r echo = FALSE}
## display selected metrics
library(knitr)
qc_f <- data.matrix(scMultiome [[c("nCount_RNA","percent.mt", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal")]])
qc_quantiles <- apply(qc_f, 2, function(x) quantile(x, probs = seq(0, 1, 0.05), na.rm = TRUE))

## select rows for displaying
qc_quantiles <- qc_quantiles[c(1, 2, 6, 11, 16, 20 , 21), ]

#rownames(sum_adj) <- c("", "Q1", "Median", "Mean", "Q3", "Max")
ti <- paste("Number of joint called cells: ", ncol(scMultiome))
kable(qc_quantiles, caption = ti)
```

### Violin plots for QC metrics
![](./`r sample_id`_VlnPlot_4_QC_metrics.pdf){width=100%, height=100%}


```{r  echo = FALSE, fig.width = 13}
## insert exsiting figures: ![](./`r sample_id`_VlnPlot_4_QC_metrics.pdf){width=100%, height=100%}
## regenerate the figure
VlnPlot(scMultiome, features = c("nCount_RNA","percent.mt", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
        ncol = 5, log = TRUE, pt.size = 0) + NoLegend()
#ggsave(paste0(out_dir, sample_id, "_VlnPlot_4_QC_metrics.pdf"), width = 12, height = 4)
```

## UMAPs
![](./`r sample_id`_UMAP_plots_clusterscMultiomeg_by_self.pdf){width=50%}

## UMAPs_2
```{r, echo=FALSE, out.width="100%", fig.cap=""}
knitr::include_graphics(paste0(sample_id, "_UMAP_plots_clustering_by_self.pdf"))
```
