<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started with Pando • Pando</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with Pando">
<meta property="og:description" content="Pando">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Pando</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting_started.html">Basic workflow</a>
    </li>
    <li>
      <a href="../articles/visualization.html">Visualization</a>
    </li>
    <li>
      <a href="../articles/regions.html">Selection of candidate regions</a>
    </li>
    <li>
      <a href="../articles/models.html">Models for GRN inference</a>
    </li>
    <li>
      <a href="../articles/association.html">Gene-peak association</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bsse.ethz.ch/qdb">
    <span class="fas fa-university fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/quadbiolab/Pando">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="getting_started_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started with Pando</h1>
                        <h4 class="author">Jonas Simon Fleck</h4>
            
            <h4 class="date">17/05/2022</h4>
      
      
      <div class="hidden name"><code>getting_started.Rmd</code></div>

    </div>

    
    
<p>This vignette gives you an introduction to the Pando workflow and gives a broad overview of the functionality. Pando is a package to infer gene regulatory networks (GRNs) from multiome data, specifically scRNA-seq and scATAC-seq. It’s is designed to interact with Seurat objects and relies on functionality from both <a href="https://satijalab.org/seurat/"><code>Seurat</code></a> and <a href="https://satijalab.org/signac/"><code>Signac</code></a>.</p>
<p>Ok, so let’s get started. First we need a Seurat object that has both RNA and ATAC modalities with gene expression and chromatin accessibility measurements, respectively. Here we use multiome data from early brain orgnaoid development</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Pando</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">muo_data</span> <span class="op">&lt;-</span> <span class="fu">read_rds</span><span class="op">(</span><span class="st">'muo_data.rds'</span><span class="op">)</span>
<span class="va">muo_data</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 173554 features across 4000 samples within 2 assays 
## Active assay: peaks (141639 features, 141639 variable features)
##  1 other assay present: RNA
##  3 dimensional reductions calculated: pca, umap, lsi</code></pre>
<p><br>
So let’s see if we got everything we need. We have an RNA assay with gene expression…</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">muo_data</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code>## Assay data with 31915 features for 4000 cells
## Top 10 variable features:
##  TAGLN, POU5F1, SST, THSD7B, AC106864.1, PAX2, MT1G, KCNIP4, FGF8, MIXL1</code></pre>
<p>…and also a <code>ChomatinAssay</code> with the ATAC data. Also we can see that we have a gene annotation already in the object. This will be important for Pando later.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">muo_data</span><span class="op">[[</span><span class="st">'peaks'</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code>## ChromatinAssay data with 141639 features for 4000 cells
## Variable features: 141639 
## Genome: 
## Annotation present: TRUE 
## Motifs present: FALSE 
## Fragment files: 0</code></pre>
<div id="initiating-the-grn" class="section level2">
<h2 class="hasAnchor">
<a href="#initiating-the-grn" class="anchor"></a>Initiating the GRN</h2>
<p>The first step in the Pando workflow is to initiate the GRN with <code><a href="../reference/initiate_grn.html">initiate_grn()</a></code>. This function also takes care of pre-selecting candidate regulatory regions to consider for GRN inference. This is optional, but in our experience constraining the set of peaks to more confident regions cuts down on runtime and makes the resulting GRN more robust. Similar to what we’ve done in <a href="https://www.biorxiv.org/content/10.1101/2021.08.24.457460v1">our preprint</a> we will here select conserved regions in mammals. The necessary data is already included in Pando:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">'phastConsElements20Mammals.UCSC.hg38'</span><span class="op">)</span>
<span class="va">muo_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/initiate_grn.html">initiate_grn</a></span><span class="op">(</span>
    <span class="va">muo_data</span>,
    rna_assay <span class="op">=</span> <span class="st">'RNA'</span>,
    peak_assay <span class="op">=</span> <span class="st">'peaks'</span>,
    regions <span class="op">=</span> <span class="va">phastConsElements20Mammals.UCSC.hg38</span> 
<span class="op">)</span></code></pre></div>
<p><br>
We have now added a <code>RegulatoryNetwork</code> object to the Seurat object. Also, you might notice that the new object is now called <code>SeuratPlus</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">muo_data</span></code></pre></div>
<pre><code>## An object of class SeuratPlus 
## 173554 features across 4000 samples within 2 assays 
## Active assay: peaks (141639 features, 141639 variable features)
##  1 other assay present: RNA
##  3 dimensional reductions calculated: pca, umap, lsi</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GetGRN.html">GetGRN</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span></code></pre></div>
<pre><code>## A RegulatoryNetwork object 
## Candidate regions have not been scanned for motifs
## 
## No network has been inferred</code></pre>
<p><br>
We can also inspect the candidate regulatory regions that we have selected:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NetworkRegions.html">NetworkRegions</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span>
<span class="va">regions</span><span class="op">@</span><span class="va">ranges</span></code></pre></div>
<pre><code>## GRanges object with 282874 ranges and 0 metadata columns:
##            seqnames            ranges strand
##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##        [1]     chr1     911275-911316      *
##        [2]     chr1     916688-916752      *
##        [3]     chr1     916755-916788      *
##        [4]     chr1     916811-916864      *
##        [5]     chr1     921178-921198      *
##        ...      ...               ...    ...
##   [282870]     chrY   6909218-6909408      *
##   [282871]     chrY 13751920-13752007      *
##   [282872]     chrY 13753848-13753884      *
##   [282873]     chrY 13803082-13803229      *
##   [282874]     chrY 14522517-14522564      *
##   -------
##   seqinfo: 308 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="scanning-for-motifs" class="section level2">
<h2 class="hasAnchor">
<a href="#scanning-for-motifs" class="anchor"></a>Scanning for motifs</h2>
<p>The next step is to scan the candidate regions for TF binding motifs, in order to have some idea where TFs might <em>potentially</em> bind. In Pando, this is done with the function <code><a href="../reference/find_motifs.html">find_motifs()</a></code>. In addition to the <code>SeuratPlus</code> object the function needs the genome and motif info in form of a <code>PFMatrixList</code>. Pando already provides a curated motif collection, but you can of course supply your own. If you choose to do so, you also need to proved a dataframe mapping motif IDs (1st column) to TF names (2nd column) with the <code>motif_tfs</code> argument. Also, if you want to infer the GRN only for a subset of TFs, you can constrain it with this function.</p>
<p>Here, we use the motifs provided by Pando, but constrain them to only contain genes involved in patterning:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">'motifs'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">'motif2tf'</span><span class="op">)</span>
<span class="va">patterning_genes</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="st">'patterning_genes.tsv'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pattern_tfs</span> <span class="op">&lt;-</span> <span class="va">patterning_genes</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">type</span><span class="op">==</span><span class="st">'Transcription factor'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">pull</span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>
<span class="va">motif2tf_use</span> <span class="op">&lt;-</span> <span class="va">motif2tf</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">tf</span> <span class="op">%in%</span> <span class="va">pattern_tfs</span><span class="op">)</span>
<span class="va">motifs_use</span> <span class="op">&lt;-</span> <span class="va">motifs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">motif2tf_use</span><span class="op">$</span><span class="va">motif</span><span class="op">)</span><span class="op">]</span>
<span class="va">motif2tf_use</span></code></pre></div>
<pre><code>## # A tibble: 216 x 8
##    motif       tf    origin gene_id         family name  symbol motif_tf
##    &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;  &lt;lgl&gt;   
##  1 M01722_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  2 M02790_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  3 M02791_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  4 M04131_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  5 M04132_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  6 M04133_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  7 M04134_2.00 HEYL  CIS-BP ENSG00000163909 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  8 M08714_2.00 TAL2  CIS-BP ENSG00000186051 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
##  9 M08734_2.00 TAL2  CIS-BP ENSG00000186051 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
## 10 M08759_2.00 TAL2  CIS-BP ENSG00000186051 bHLH   &lt;NA&gt;  &lt;NA&gt;   NA      
## # … with 206 more rows</code></pre>
<p><br>
Using these TF motifs, we can now scan the candidate regions:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span>
<span class="va">muo_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_motifs.html">find_motifs</a></span><span class="op">(</span>
    <span class="va">muo_data</span>, 
    pfm <span class="op">=</span> <span class="va">motifs_use</span>, 
    motif_tfs <span class="op">=</span> <span class="va">motif2tf_use</span>,
    genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>
<span class="op">)</span></code></pre></div>
<p><br>
Our <code>Regions</code> object has now gotten a new slot containing the <code>Motif</code> object. This object stores a sparse peak x motif matrix with the matches and some other information.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NetworkRegions.html">NetworkRegions</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span>
<span class="va">regions</span><span class="op">@</span><span class="va">motifs</span><span class="op">@</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code>## 5 x 5 sparse Matrix of class "dgCMatrix"
##                    M01722_2.00 M02790_2.00 M02791_2.00 M04131_2.00 M04132_2.00
## chr1-911275-911316           .           .           .           .           .
## chr1-916688-916752           .           .           .           .           .
## chr1-916755-916788           .           .           .           .           .
## chr1-916811-916864           .           .           .           .           .
## chr1-921178-921198           .           .           .           .           .</code></pre>
</div>
<div id="inferring-the-grn" class="section level2">
<h2 class="hasAnchor">
<a href="#inferring-the-grn" class="anchor"></a>Inferring the GRN</h2>
<p>Now we should have everything ready to infer the GRN. GRN inference in Pando is very much inspired by other methods such as <a href="https://scenic.aertslab.org/">SCENIC</a>. These methods assume that the expression of a gene can be modeled by a function of the expression of the TFs that regulate it. Pando extends this notion to harness multiome measurements by considering co-accessibility in addition to co-expression. The underlying idea is that in order for a TF to regulate a gene, the TF needs to be expressed <strong>and</strong> the binding peak needs to be accessible. For this, Pando fits one regression model per gene, and models its expression based on TF expression and accessibility of the binding peaks. These TF-peak interactions are additively combined. If you are familiar with the R formula syntax, you might recognize something like this:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">PAX6</span> <span class="op">~</span> <span class="va">HES4</span><span class="op">*</span><span class="va">chr1</span><span class="op">-</span><span class="fl">911275</span><span class="op">-</span><span class="fl">911316</span> <span class="op">+</span> <span class="va">OTX2</span><span class="op">*</span><span class="va">chr1</span><span class="op">-</span><span class="fl">921178</span><span class="op">-</span><span class="fl">921198</span> <span class="op">+</span> <span class="va">MSX2</span><span class="op">*</span><span class="va">chr1</span><span class="op">-</span><span class="fl">921178</span><span class="op">-</span><span class="fl">921198</span> <span class="op">+</span> <span class="va">...</span></code></pre></div>
<p>In this example, PAX6 expression is modeled by adding up the TF-peak interactions of HES4, OTX2 and MSX2.</p>
<p>The function <code><a href="../reference/infer_grn.html">infer_grn()</a></code> takes care of fitting these models for all genes and identifies significant connections between TF-region pairs and target gene expression. With the <code>genes</code> argument, we can select a subset of genes that we want to use for GRN inference. Here we stick with the theme above and use pattering-related genes. This is of course optional, but for runtime reasons we recommend to constrain the set of genes somehow, e.g. by using <code>VariableFeatures()</code>. Also, we use the method used by <a href="http://great.stanford.edu/public/html/">GREAT</a> to associate genes with peaks and parallelize the computation using 4 cores:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">doParallel</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html">registerDoParallel</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span>
<span class="va">muo_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_grn.html">infer_grn</a></span><span class="op">(</span>
    <span class="va">muo_data</span>,
    peak_to_gene_method <span class="op">=</span> <span class="st">'GREAT'</span>,
    genes <span class="op">=</span> <span class="va">patterning_genes</span><span class="op">$</span><span class="va">symbol</span>,
    parallel <span class="op">=</span> <span class="cn">T</span>
<span class="op">)</span></code></pre></div>
<p><code><a href="../reference/infer_grn.html">infer_grn()</a></code> has many more parameters to choose different models or tweak the association between peaks and genes, but we’ll go into this in another vignette.</p>
<p>The inferred <code>Network</code> object can be accessed with <code><a href="../reference/GetNetwork.html">GetNetwork()</a></code></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GetNetwork.html">GetNetwork</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span></code></pre></div>
<pre><code>## A Network object
## with 57 TFs and 122 target genes</code></pre>
<p>and the parameters of the models can be inspected with <code><a href="https://rdrr.io/r/stats/coef.html">coef()</a></code>. This returns a dataframe coefficients (<code>estimate</code>) and p-values (<code>pval</code>) for each TF/region-target gene pair:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3,804 x 10
##    tf     target region  term   estimate std_err statistic    pval   padj   corr
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
##  1 DMRTA1 IRS4   chrX-1… chrX_…  0.0417   0.0157    2.65   0.00801 0.0440  0.128
##  2 DMRTA1 IRS4   chrX-1… DMRTA…  0.0155   0.0184    0.844  0.399   0.671   0.128
##  3 DMRTA1 IRS4   chrX-1… DMRTA…  0.0499   0.0237    2.10   0.0354  0.138   0.128
##  4 BNC2   FGF13  chrX-1… chrX_… -0.0111   0.0214   -0.521  0.602   0.814  -0.143
##  5 BNC2   FGF13  chrX-1… BNC2:… -0.0110   0.0237   -0.465  0.642   0.840  -0.143
##  6 ID3    FGF13  chrX-1… chrX_… -0.0181   0.0220   -0.821  0.412   0.685   0.125
##  7 ID3    FGF13  chrX-1… ID3:c…  0.0519   0.0477    1.09   0.277   0.555   0.125
##  8 EMX2   FGF13  chrX-1… chrX_…  0.00191  0.0549    0.0347 0.972   0.993   0.120
##  9 OTX2   FGF13  chrX-1… chrX_…  0.0844   0.0342    2.47   0.0137  0.0669  0.106
## 10 ID3    FGF13  chrX-1… ID3:c…  0.00820  0.0220    0.373  0.709   0.879   0.125
## # … with 3,794 more rows</code></pre>
</div>
<div id="module-discovery" class="section level2">
<h2 class="hasAnchor">
<a href="#module-discovery" class="anchor"></a>Module discovery</h2>
<p>Using the inferred parameters, we can now use <code><a href="../reference/find_modules.html">find_modules()</a></code> to construct TF modules, i.e. the set of genes that are regulated by each transcription factor. How exactly this is done depends on the model choice in <code><a href="../reference/infer_grn.html">infer_grn()</a></code>, but generally Pando chooses the most confident connections for each TF. The function allows setting different selection criteria like a p-value threshold or an <span class="math inline">\(R^2\)</span> threshold to filter models by goodness-of-fit. For the purpose of this tutorial we will choose rather lenient thresholds:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">muo_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_modules.html">find_modules</a></span><span class="op">(</span>
    <span class="va">muo_data</span>, 
    p_thresh <span class="op">=</span> <span class="fl">0.1</span>,
    nvar_thresh <span class="op">=</span> <span class="fl">2</span>, 
    min_genes_per_module <span class="op">=</span> <span class="fl">1</span>, 
    rsq_thresh <span class="op">=</span> <span class="fl">0.05</span>
<span class="op">)</span></code></pre></div>
<p><br>
The <code>Modules</code> object can be accessed with <code><a href="../reference/NetworkModules.html">NetworkModules()</a></code>. The <code>meta</code> slot contains meta data about the module network and lists with feature sets for each TF can be accessed in the <code>features</code> slot.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NetworkModules.html">NetworkModules</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span> 
<span class="va">modules</span><span class="op">@</span><span class="va">meta</span></code></pre></div>
<pre><code>## # A tibble: 330 x 9
## # Groups:   target, tf [330]
##    tf    target estimate n_regions n_genes n_tfs regions           pval     padj
##    &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;     &lt;int&gt;   &lt;int&gt; &lt;int&gt; &lt;chr&gt;            &lt;dbl&gt;    &lt;dbl&gt;
##  1 BNC2  BAMBI    0.395         13      28    10 chr10-285331… 1.57e-18 5.26e-17
##  2 BNC2  BMP4     0.636         18      28     9 chr14-535716… 2.37e-11 5.94e-10
##  3 BNC2  BMP5     0.327          8      28     8 chr6-5569415… 2.60e-24 1.01e-22
##  4 BNC2  BMPER    0.450         17      28    13 chr7-3345788… 1.59e- 9 3.48e- 8
##  5 BNC2  DMRTA1  -0.110          6      28     6 chr9-2180250… 1.77e- 5 2.19e- 4
##  6 BNC2  EMX2    -0.112         12      28    10 chr10-117139… 4.72e- 3 2.91e- 2
##  7 BNC2  FEZF2   -0.0670         7      28     6 chr3-6287628… 1.24e- 3 9.82e- 3
##  8 BNC2  FOS      0.0681         7      28     6 chr14-759824… 8.14e- 3 4.45e- 2
##  9 BNC2  FOXF1    0.130          9      28     4 chr16-864943… 2.52e-10 5.90e- 9
## 10 BNC2  HOXB1    0.433         24      28     7 chr17-482253… 2.24e- 7 3.90e- 6
## # … with 320 more rows</code></pre>
<p><br>
The goodness-of-fit metrics can be plotted with</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_gof.html">plot_gof</a></span><span class="op">(</span><span class="va">muo_data</span>, point_size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="getting_started_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>and the size of the modules can be plotted with</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_module_metrics.html">plot_module_metrics</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span></code></pre></div>
<p><img src="getting_started_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
<div id="visualizing-the-grn" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-the-grn" class="anchor"></a>Visualizing the GRN</h2>
<p>Finally, we can vizualize the GRN. For this, we first need to create the graph to be visualized and optionally a UMAP embedding for the nodes with the function <code><a href="../reference/get_network_graph.html">get_network_graph()</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">muo_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_network_graph.html">get_network_graph</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span></code></pre></div>
<p>Then we can plot the graph with <code><a href="../reference/plot_network_graph.html">plot_network_graph()</a></code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_network_graph.html">plot_network_graph</a></span><span class="op">(</span><span class="va">muo_data</span><span class="op">)</span></code></pre></div>
<p><img src="getting_started_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>The default here is to use a UMAP embedding, but you can choose any layout option provided by <a href="https://igraph.org/"><code>igraph</code></a>/<a href="https://ggraph.data-imaginist.com/index.html"><code>ggraph</code></a>, for instance the force-directed Fruchterman–Reingold layout:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_network_graph.html">plot_network_graph</a></span><span class="op">(</span><span class="va">muo_data</span>, layout<span class="op">=</span><span class="st">'fr'</span><span class="op">)</span></code></pre></div>
<p><img src="getting_started_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jonas Simon Fleck.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
